packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/FileSelectorDelegate.java

```diff
@@ -0,0 +1,217 @@
+    if (setPendingResult(result)) {
+      finishWithAlreadyActiveError(result);
+      return;
+    }
+
+    Boolean multipleFiles = options.getAllowMultiple();
+    List<String> acceptedTypeGroups = options.getAllowedTypes();
+
+    launchOpenFile(multipleFiles, acceptedTypeGroups);
+  }
+
+  void launchGetDirectoryPath(@Nullable String initialDirectory) {
+    Intent getDirectoryPathIntent = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
+
+    if (initialDirectory != null && !initialDirectory.isEmpty()) {
+      Uri uri = getDirectoryPathIntent.getParcelableExtra("android.provider.extra.INITIAL_URI");
+      String scheme = uri.toString();
+      scheme = scheme.replace("/root/", initialDirectory);
+      uri = Uri.parse(scheme);
+      getDirectoryPathIntent.putExtra("android.provider.extra.INITIAL_URI", uri);
+    }
+    activity.startActivityForResult(getDirectoryPathIntent, REQUEST_CODE_GET_DIRECTORY_PATH);
+  }
+
+  @Override
+  public boolean onActivityResult(int requestCode, int resultCode, Intent data) {
+    switch (requestCode) {
+      case REQUEST_CODE_GET_DIRECTORY_PATH:
+        handleGetDirectoryPathResult(resultCode, data);
+        break;
+      case REQUEST_CODE_OPEN_FILE:
+        handleOpenFileResult(resultCode, data);
+        break;
+      default:
+        return false;
+    }
+
+    return true;
+  }
+
+  void launchOpenFile(boolean isMultipleSelection, List<String> acceptedTypeGroups) {
+    openFileIntent.setAction(Intent.ACTION_GET_CONTENT);
+    openFileIntent.addCategory(Intent.CATEGORY_OPENABLE);
+    openFileIntent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, isMultipleSelection);
+
+    if (acceptedTypeGroups != null && !acceptedTypeGroups.isEmpty()) {
+      openFileIntent.setType(acceptedTypeGroups.get(0));
+
+      openFileIntent.putExtra(Intent.EXTRA_MIME_TYPES, acceptedTypeGroups.toArray(new String[0]));
+    }
+
+    activity.startActivityForResult(openFileIntent, REQUEST_CODE_OPEN_FILE);
+  }
+
+  private void handleGetDirectoryPathResult(int resultCode, Intent data) {
+    if (resultCode == Activity.RESULT_OK && data != null) {
+      Uri path = data.getData();
+      handleGetDirectoryPathResult(path.toString());
+      return;
+    }
+
+    finishWithSuccess(null);
+  }
+
+  void handleOpenFileResult(int resultCode, Intent data) {
+    if (resultCode != Activity.RESULT_OK || data == null) {
+      finishWithSuccess(null);
+      return;
+    }
+
+    ArrayList<Uri> uris = uriHandler(data);
+
+    ArrayList<String> srcPaths =
+        PathUtils.copyFilesToInternalStorage(uris, this.activity, cacheFolder);
+    handleOpenFileActionResults(srcPaths);
+  }
+
+  private void handleGetDirectoryPathResult(String path) {
+    finishWithSuccess(path);
+  }
+
+  void handleOpenFileActionResults(ArrayList<String> srcPaths) {
+    finishWithListSuccess(srcPaths);
+  }
+
+  private <T> boolean setPendingResult(Messages.Result<T> result) {
```

Comment:
This needs a declaration comment explaining what it does and returns, especially since the return value is non-obvious.

Reply:
I'm rewriting this.